src/Akka.Streams.Kafka/Stages/Consumers/Actors/KafkaConsumerActor.cs

```diff
@@ -556,6 +609,33 @@ private void Poll()
                 using (var cts = new CancellationTokenSource(_settings.PollTimeout))
                 {
                     var (polled, exception) = PollKafka(cts.Token);
+                    var assignedAtEnd = _consumer.Assignment.ToImmutableHashSet();
+                    
+                    /*
+                     * Any partitions that were NOT ASSIGNED at the start of the poll BUT ARE ASSIGNED NOW
+                     * got assigned to us mid-poll then. We need to stash these messages until they are requested.
+                     */
+                    var newlyAssigned = assignedAtEnd.Except(currentAssignment);
+                    if (newlyAssigned.Any())
+                    {
+                        // now filter again to see if any of these newlyAssigned partitions have not been requested
+                        var newlyAssignedButNotRequested = newlyAssigned.Except(partitionsToFetch);
+                        if (newlyAssignedButNotRequested.Any())
+                        {
+                            var (newUnrequested, requested) = polled.Partition(c => newlyAssignedButNotRequested.Contains(c.TopicPartition));
+                            if (newUnrequested.Count > 0)
+                            {
+                                var originalUnrequestedCount = _unRequestedMessages.Count;
```

Comment:
If we've made it here, we have found the edge case that triggered #415 previously - a newly assigned partition that has not yet been requested which ALSO served up a message immediately to us during our poll attempt.

We're going to filter the set of messages we received back and ONLY transmit the `ConsumeResult<K,V>` belonging to the `_requests` to the `ProcessResult` method. All of the other pre-processing we do here is to just stash the messages that we retrieved but haven't received a request for yet. Hopefully the next round of `RequestMessages` will satisfy that.